File: src/main/java/itmo/is/project/service/module/StorageModuleService.java
package itmo.is.project.service.module;

import itmo.is.project.dto.module.storage.StorageModuleDto;
import itmo.is.project.dto.resource.ResourceAmountDto;
import itmo.is.project.dto.module.storage.StoredResourceDto;
import itmo.is.project.mapper.module.storage.StorageModuleMapper;
import itmo.is.project.mapper.resource.ResourceAmountMapper;
import itmo.is.project.mapper.module.storage.StoredResourceMapper;
import itmo.is.project.model.module.storage.StorageModuleFreeSpace;
import itmo.is.project.model.module.storage.StoredResource;
import itmo.is.project.model.resource.ResourceAmount;
import itmo.is.project.model.resource.ResourceAmountHolder;
import itmo.is.project.model.resource.ResourceIdAmountHolder;
import itmo.is.project.repository.module.storage.StorageModuleRepository;
import itmo.is.project.repository.module.storage.StoredResourceRepository;
import itmo.is.project.service.resource.ResourceService;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Service
@RequiredArgsConstructor
public class StorageModuleService {

    private final StorageModuleRepository storageModuleRepository;
    private final StorageModuleMapper storageModuleMapper;

    private final StoredResourceRepository storedResourceRepository;
    private final StoredResourceMapper storedResourceMapper;
    private final ResourceAmountMapper resourceAmountMapper;

    private final ResourceService resourceService;

    public Page<StorageModuleDto> getAllStorageModules(Pageable pageable) {
        return storageModuleRepository.findAll(pageable).map(storageModuleMapper::toDto);
    }

    public StorageModuleDto getStorageModuleById(Integer id) {
        return storageModuleRepository.findById(id).map(storageModuleMapper::toDto)
                .orElseThrow(() -> new EntityNotFoundException("Storage module not found with id: " + id));
    }

    public Page<StoredResourceDto> getAllStoredResources(Pageable pageable) {
        return storedResourceRepository.findAll(pageable)
                .map(storedResourceMapper::toDto);
    }

    public Page<ResourceAmountDto> getAllResourcesTotal(Pageable pageable) {
        return storedResourceRepository.findAllResourceAmountsTotal(pageable)
                .map(resourceAmountMapper::toDto);
    }

    public ResourceAmountDto getResourceAmountTotalByResourceId(Integer resourceId) {
        return storedResourceRepository.findResourceAmountTotal(resourceId)
                .map(resourceAmountMapper::toDto)
                .orElseThrow(() ->
                        new EntityNotFoundException("Resource not found in storages with resource id: " + resourceId)
                );
    }

    public Page<ResourceAmountDto> getAllResourcesByStorageId(Integer storageModuleId, Pageable pageable) {
        return storedResourceRepository.findAllByIdStorageModuleId(storageModuleId, pageable)
                .map(resourceAmountMapper::toDto);
    }


    @Transactional
    public void storeById(ResourceIdAmountHolder resourceIdAmount) {
        store(resourceService.toResourceAmount(resourceIdAmount));
    }

    @Transactional
    public void store(ResourceAmountHolder resourceAmount) {
        storeAll(List.of(resourceAmount));
    }

    @Transactional
    public void storeAllById(Collection<? extends ResourceIdAmountHolder> resources) {
        storeAll(resources.stream().map(resourceService::toResourceAmount).toList());
    }

    @Transactional
    public void storeAll(Collection<? extends ResourceAmountHolder> resources) {
        int amountTotal = sumResourcesAmount(resources);
        checkFreeSpace(amountTotal);
        Deque<StorageModuleFreeSpace> storages = getAvailableStorages();
        for (ResourceAmountHolder resourceAmount : resources) {
            storeResourceToStorages(resourceAmount, storages);
        }
    }

    private Deque<StorageModuleFreeSpace> getAvailableStorages() {
        return new ArrayDeque<>(storageModuleRepository.findAllHavingFreeSpace());
    }

    private void storeResourceToStorages(
            ResourceAmountHolder resourceAmount,
            Deque<StorageModuleFreeSpace> storages
    ) {
        int remainingAmount = resourceAmount.getAmount();
        while (remainingAmount > 0) {
            StorageModuleFreeSpace storageModuleFreeSpace = storages.poll();
            remainingAmount -= storeResourceUpToStorageCapacity(resourceAmount, storageModuleFreeSpace);
            if (storageModuleFreeSpace.getFreeSpace() > 0) {
                storages.offerFirst(storageModuleFreeSpace);
            }
        }
    }

    private int storeResourceUpToStorageCapacity(
            ResourceAmountHolder resourceAmountToStore,
            StorageModuleFreeSpace storageModuleFreeSpace
    ) {
        StoredResource.CompositeKey id = new StoredResource.CompositeKey(
                storageModuleFreeSpace.getStorageModuleId(), resourceAmountToStore.getResourceId()
        );
        StoredResource storedResource = storedResourceRepository.findById(id).orElseGet(() ->
                new StoredResource(id, storageModuleFreeSpace.getStorageModule(), resourceAmountToStore.getResource(), 0)
        );
        int storeAmount = Math.min(resourceAmountToStore.getAmount(), storageModuleFreeSpace.getFreeSpace());
        storedResource.add(storeAmount);
        storedResourceRepository.save(storedResource);
        storageModuleFreeSpace.sub(storeAmount);
        return storeAmount;
    }


    @Transactional
    public void retrieveById(ResourceIdAmountHolder resourceIdAmount) {
        retrieve(resourceService.toResourceAmount(resourceIdAmount));
    }

    @Transactional
    public void retrieve(ResourceAmountHolder resourceAmount) {
        retrieveAll(List.of(resourceAmount));
    }

    @Transactional
    public void retrieveAllById(Collection<? extends ResourceIdAmountHolder> resources) {
        retrieveAll(resources.stream().map(resourceService::toResourceAmount).toList());
    }

    @Transactional
    public void retrieveAll(Collection<? extends ResourceAmountHolder> resources) {
        for (ResourceAmountHolder resourceAmount : resources) {
            checkExistenceRequiredResourceAmount(resourceAmount);
        }
        for (ResourceAmountHolder resourceAmount : resources) {
            retrieveResourceFromStorages(resourceAmount);
        }
    }

    private void checkExistenceRequiredResourceAmount(ResourceIdAmountHolder required) {
        Integer resourceId = required.getResourceId();
        ResourceAmount total = storedResourceRepository.findResourceAmountTotal(resourceId).orElseThrow();
        if (total.getAmount() < required.getAmount()) {
            throw new IllegalStateException("Insufficient resources");
        }
    }

    private void retrieveResourceFromStorages(ResourceIdAmountHolder resourceIdAmount) {
        Integer resourceId = resourceIdAmount.getResourceId();
        List<StoredResource> storedResources = storedResourceRepository.findAllByIdResourceId(resourceId);

        int remainingAmount = resourceIdAmount.getAmount();
        for (StoredResource storedResource : storedResources) {
            if (storedResource.getAmount() <= remainingAmount) {
                storedResourceRepository.delete(storedResource);
                remainingAmount -= storedResource.getAmount();
            } else {
                storedResource.setAmount(storedResource.getAmount() - remainingAmount);
                storedResourceRepository.save(storedResource);
                remainingAmount = 0;
            }
            if (remainingAmount == 0) {
                break;
            }
        }
    }


    @Transactional
    public void retrieveAndStoreAllById(
            Collection<? extends ResourceIdAmountHolder> retrieve,
            Collection<? extends ResourceIdAmountHolder> store
    ) {
        retrieveAndStoreAll(
                retrieve.stream().map(resourceService::toResourceAmount).toList(),
                store.stream().map(resourceService::toResourceAmount).toList()
        );
    }

    @Transactional
    public void retrieveAndStoreAll(
            Collection<? extends ResourceAmountHolder> retrieve,
            Collection<? extends ResourceAmountHolder> store
    ) {
        int requiredSpace = sumResourcesAmount(store) - sumResourcesAmount(retrieve);
        if (requiredSpace > 0) {
            checkFreeSpace(requiredSpace);
        }
        retrieveAll(retrieve);
        storeAll(store);
    }


    private int sumResourcesAmount(Collection<? extends ResourceIdAmountHolder> resources) {
        return resources.stream()
                .mapToInt(ResourceIdAmountHolder::getAmount)
                .sum();
    }

    private void checkFreeSpace(int amount) {
        int freeSpace = storageModuleRepository.getTotalFreeSpaceInStorages();
        if (amount > freeSpace) {
            throw new IllegalStateException("Not enough free space in storages");
        }
    }
}

File: src/main/kotlin/itmo/isproject/service/module/StorageModuleService.kt
package itmo.isproject.service.module

import itmo.isproject.dto.module.storage.StorageModuleDto
import itmo.isproject.dto.resource.ResourceAmountDto
import itmo.isproject.dto.module.storage.StoredResourceDto
import io.github.oshai.kotlinlogging.KotlinLogging
import io.github.oshai.kotlinlogging.withLoggingContext
import itmo.isproject.mapper.module.storage.StorageModuleMapper
import itmo.isproject.mapper.resource.ResourceAmountMapper
import itmo.isproject.mapper.module.storage.StoredResourceMapper
import itmo.isproject.model.module.storage.StorageModuleFreeSpace
import itmo.isproject.model.module.storage.StoredResource
import itmo.isproject.model.module.storage.StoredResourceId
import itmo.isproject.model.resource.ResourceAmountHolder
import itmo.isproject.model.resource.ResourceIdAmountHolder
import itmo.isproject.repository.module.storage.StorageModuleRepository
import itmo.isproject.repository.module.storage.StoredResourceRepository
import itmo.isproject.service.resource.ResourceService
import jakarta.persistence.EntityNotFoundException
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

private val logger = KotlinLogging.logger {}

@Service
class StorageModuleService(
    private val storageModuleRepository: StorageModuleRepository,
    private val storageModuleMapper: StorageModuleMapper,
    private val storedResourceRepository: StoredResourceRepository,
    private val storedResourceMapper: StoredResourceMapper,
    private val resourceAmountMapper: ResourceAmountMapper,
    private val resourceService: ResourceService
) {

    fun getAllStorageModules(pageable: Pageable): Page<StorageModuleDto> {
        withLoggingContext("page" to pageable.pageNumber.toString()) {
            logger.debug { "Fetching all storage modules" }
        }
        return storageModuleRepository.findAll(pageable).map { storageModuleMapper.toDto(it) }
    }

    fun getStorageModuleById(id: Int): StorageModuleDto {
        withLoggingContext("storageModuleId" to id.toString()) {
            logger.debug { "Fetching storage module by" }
        }
        return storageModuleRepository.findByIdOrNull(id)?.let { storageModuleMapper.toDto(it) }
            ?: throw EntityNotFoundException("Storage module not found with id: $id")
    }

    fun getAllStoredResources(pageable: Pageable): Page<StoredResourceDto> {
        withLoggingContext("page" to pageable.pageNumber.toString()) {
            logger.debug { "Fetching all stored resources" }
        }
        return storedResourceRepository.findAll(pageable)
            .map { storedResourceMapper.toDto(it) }
    }

    fun getAllResourcesTotal(pageable: Pageable): Page<ResourceAmountDto> {
        withLoggingContext("page" to pageable.pageNumber.toString()) {
            logger.debug { "Fetching all resources total" }
        }
        return storedResourceRepository.findAllResourceAmountsTotal(pageable)
            .map { resourceAmountMapper.toDto(it) }
    }

    fun getResourceAmountTotalByResourceId(resourceId: Int): ResourceAmountDto {
        withLoggingContext("resourceId" to resourceId.toString()) {
            logger.debug { "Fetching resource amount total by" }
        }
        return storedResourceRepository.findResourceAmountTotal(resourceId)
            ?.let { resourceAmountMapper.toDto(it) }
            ?: throw EntityNotFoundException("Resource not found in storages with resource id: $resourceId")
    }

    fun getAllResourcesByStorageId(storageModuleId: Int, pageable: Pageable): Page<ResourceAmountDto> {
        withLoggingContext("storageModuleId" to storageModuleId.toString(), "page" to pageable.pageNumber.toString()) {
            logger.debug { "Fetching all resources by" }
        }
        return storedResourceRepository.findAllByIdStorageModuleId(storageModuleId, pageable)
            .map { resourceAmountMapper.toDto(it) }
    }

    @Transactional
    fun storeById(resourceIdAmount: ResourceIdAmountHolder) {
        store(resourceService.toResourceAmount(resourceIdAmount))
    }

    @Transactional
    fun store(resourceAmount: ResourceAmountHolder) {
        storeAll(listOf(resourceAmount))
    }

    @Transactional
    fun storeAllById(resources: Collection<ResourceIdAmountHolder>) {
        storeAll(resources.map { resourceService.toResourceAmount(it) })
    }

    @Transactional
    fun storeAll(resources: Collection<ResourceAmountHolder>) {
        withLoggingContext("resourcesCount" to resources.size.toString()) {
            logger.info { "Storing resources" }
        }
        val amountTotal = sumResourcesAmount(resources)
        withLoggingContext("amount" to amountTotal.toString()) {
            logger.debug { "Total amount to store" }
        }
        checkFreeSpace(amountTotal)
        val storages = getAvailableStorages()
        for (resourceAmount in resources) {
            storeResourceToStorages(resourceAmount, storages)
        }
        withLoggingContext("resourcesCount" to resources.size.toString()) {
            logger.info { "Resources stored successfully" }
        }
    }

    private fun getAvailableStorages(): Deque<StorageModuleFreeSpace> {
        return ArrayDeque(storageModuleRepository.findAllHavingFreeSpace())
    }

    private fun storeResourceToStorages(
        resourceAmount: ResourceAmountHolder,
        storages: Deque<StorageModuleFreeSpace>
    ) {
        var remainingAmount = resourceAmount.amount ?: 0
        while (remainingAmount > 0) {
            val storageModuleFreeSpace = storages.poll()
            remainingAmount -= storeResourceUpToStorageCapacity(resourceAmount, storageModuleFreeSpace)
            if (storageModuleFreeSpace.freeSpace > 0) {
                storages.offerFirst(storageModuleFreeSpace)
            }
        }
    }

    private fun storeResourceUpToStorageCapacity(
        resourceAmountToStore: ResourceAmountHolder,
        storageModuleFreeSpace: StorageModuleFreeSpace
    ): Int {
        val storageModule = storageModuleFreeSpace.storageModule!!
        val storageModuleId = storageModule.id
        val resourceId = resourceAmountToStore.resourceId!!
        val storedResourceId = StoredResourceId(storageModuleId, resourceId)
        val storedResource = storedResourceRepository.findByIdOrNull(storedResourceId) ?: StoredResource(
            resourceAmountToStore,
            storageModule
        )
        val storeAmount = (resourceAmountToStore.amount ?: 0).coerceAtMost(storageModuleFreeSpace.freeSpace)
        storedResource += storeAmount
        storedResourceRepository.save(storedResource)
        storageModuleFreeSpace -= storeAmount
        return storeAmount
    }

    @Transactional
    fun retrieveById(resourceIdAmount: ResourceIdAmountHolder) {
        retrieve(resourceService.toResourceAmount(resourceIdAmount))
    }

    @Transactional
    fun retrieve(resourceAmount: ResourceAmountHolder) {
        retrieveAll(listOf(resourceAmount))
    }

    @Transactional
    fun retrieveAllById(resources: Collection<ResourceIdAmountHolder>) {
        retrieveAll(resources.map { resourceService.toResourceAmount(it) })
    }

    @Transactional
    fun retrieveAll(resources: Collection<ResourceAmountHolder>) {
        withLoggingContext("resourcesCount" to resources.size.toString()) {
            logger.info { "Retrieving resources" }
        }
        for (resourceAmount in resources) {
            checkExistenceRequiredResourceAmount(resourceAmount)
        }
        for (resourceAmount in resources) {
            retrieveResourceFromStorages(resourceAmount)
        }
        withLoggingContext("resourcesCount" to resources.size.toString()) {
            logger.info { "Resources retrieved successfully" }
        }
    }

    private fun checkExistenceRequiredResourceAmount(required: ResourceIdAmountHolder) {
        val resourceId = required.resourceId
        val total = storedResourceRepository.findResourceAmountTotal(resourceId)
            ?: throw EntityNotFoundException("Resource not found in storages with resource id: $resourceId")
        if ((total.amount ?: 0) < (required.amount ?: 0)) {
            withLoggingContext(
                "resourceId" to (resourceId?.toString() ?: "null"),
                "required" to (required.amount?.toString() ?: "null"),
                "available" to (total.amount?.toString() ?: "null")
            ) {
                logger.warn { "Insufficient resources" }
            }
            throw IllegalStateException("Insufficient resources")
        }
    }

    private fun retrieveResourceFromStorages(resourceIdAmount: ResourceIdAmountHolder) {
        val resourceId = resourceIdAmount.resourceId
        val storedResources = storedResourceRepository.findAllByIdResourceId(resourceId)

        var remainingAmount = resourceIdAmount.amount ?: 0
        for (storedResource in storedResources) {
            if ((storedResource.amount ?: 0) <= remainingAmount) {
                storedResourceRepository.delete(storedResource)
                remainingAmount -= storedResource.amount ?: 0
            } else {
                storedResource.amount = (storedResource.amount ?: 0) - remainingAmount
                storedResourceRepository.save(storedResource)
                remainingAmount = 0
            }
            if (remainingAmount == 0) {
                break
            }
        }
    }

    @Transactional
    fun retrieveAndStoreAllById(
        retrieve: Collection<ResourceIdAmountHolder>,
        store: Collection<ResourceIdAmountHolder>
    ) {
        retrieveAndStoreAll(
            retrieve.map { resourceService.toResourceAmount(it) },
            store.map { resourceService.toResourceAmount(it) }
        )
    }

    @Transactional
    fun retrieveAndStoreAll(
        retrieve: Collection<ResourceAmountHolder>,
        store: Collection<ResourceAmountHolder>
    ) {
        val requiredSpace = sumResourcesAmount(store) - sumResourcesAmount(retrieve)
        if (requiredSpace > 0) {
            checkFreeSpace(requiredSpace)
        }
        retrieveAll(retrieve)
        storeAll(store)
    }

    private fun sumResourcesAmount(resources: Collection<ResourceIdAmountHolder>): Int {
        return resources.sumOf { it.amount ?: 0 }
    }

    private fun checkFreeSpace(amount: Int) {
        val freeSpace = storageModuleRepository.getTotalFreeSpaceInStorages() ?: 0
        withLoggingContext("required" to amount.toString(), "available" to freeSpace.toString()) {
            logger.debug { "Checking free space" }
        }
        if (amount > freeSpace) {
            withLoggingContext("required" to amount.toString(), "available" to freeSpace.toString()) {
                logger.warn { "Not enough free space" }
            }
            throw IllegalStateException("Not enough free space in storages")
        }
    }
}
